logins-sql/src/engine.rs

```diff
@@ -0,0 +1,296 @@
+        self.db.reset()
+    }
+
+    pub fn update(&self, login: Login) -> Result<()> {
+        self.db.update(login)
+    }
+
+    pub fn add(&self, login: Login) -> Result<String> {
+        // Just return the record's ID (which we may have generated).
+        self.db.add(login).map(|record| record.id)
+    }
+
+    // This is basiclaly exposed just for sync_pass_sql, but it doesn't seem
+    // unreasonable.
+    pub fn conn(&self) -> &rusqlite::Connection {
+        &self.db.db
+    }
+
+    pub fn sync(
+        &mut self,
+        storage_init: &Sync15StorageClientInit,
+        root_sync_key: &KeyBundle
+    ) -> Result<()> {
+
+        // Note: If `to_ready` (or anything else with a ?) fails below, this
+        // `take()` means we end up with `state.sync.is_none()`, which means the
+        // next sync will redownload meta/global, crypto/keys, etc. without
+        // needing to. Apparently this is both okay and by design.
+        let maybe_sync_info = self.sync.take().map(Ok);
+
+        // `maybe_sync_info` is None if we haven't called `sync` since
+        // restarting the browser.
+        //
+        // If this is the case we may or may not have a persisted version of
+        // GlobalState stored in the DB (we will iff we've synced before, unless
+        // we've `reset()`, which clears it out).
+        let mut sync_info = maybe_sync_info.unwrap_or_else(|| -> Result<SyncInfo> {
+            info!("First time through since unlock. Trying to load persisted global state.");
+            let state = if let Some(persisted_global_state) = self.db.get_global_state()? {
+                serde_json::from_str::<GlobalState>(&persisted_global_state)
+                .unwrap_or_else(|_| {
+                    // Don't log the error since it might contain sensitive
+                    // info like keys (the JSON does, after all).
+                    error!("Failed to parse GlobalState from JSON! Falling back to default");
+                    // Unstick ourselves by using the default state.
+                    GlobalState::default()
+                })
+            } else {
+                info!("No previously persisted global state, using default");
+                GlobalState::default()
+            };
+            let client = Sync15StorageClient::new(storage_init.clone())?;
+            Ok(SyncInfo {
+                state,
+                client,
+                last_client_init: storage_init.clone(),
+            })
+        })?;
+
+        // If the options passed for initialization of the storage client aren't
+        // the same as the ones we used last time, reinitialize it. (Note that
+        // we could avoid the comparison in the case where we had `None` in
+        // `state.sync` before, but this probably doesn't matter).
+        //
+        // It's a little confusing that we do things this way (transparently
+        // re-initialize the client), but it reduces the size of the API surface
```

Comment:
That all makes sense, and it's nice not to expose this complexity to the consumer. I think it'd be nice to have the state machine handle this, too, and give you a fully initialized client back, since the first sync after a node reassignment is going to involve fetching (or uploading) a fresh `m/g` and `c/k`.